### PR TITLE
Add Go (Golang) language support to func init via the Native worker runtime

### PR DESCRIPTION
**Pull request by harshivcodes**
This PR adds Go language support to the `func init` command in Azure Functions Core Tools. Go is supported under the Native worker runtime, which hosts language-native binaries directly. When a user runs `func init --worker-runtime native`, the CLI scaffolds a complete Go project with module initialization, dependency resolution, and a starter `main.go`.

Core runtime registration
- Added WorkerRuntime.Native enum value with "go" and "golang" as aliases
- --worker-runtime native, --worker-runtime go, and --worker-runtime golang all resolve to Native
- When go or golang is used, a warning is printed clarifying that Golang runs under the native runtime
- Added Constants.Languages.Golang and wired up default language/programming model mappings
- Native appears in the interactive runtime selection picker

Go project scaffolding (GoHelpers.cs)
- Detects installed Go version (minimum 1.24 required)
- Runs go mod init, go get github.com/azure/azure-functions-golang-worker, and go mod tidy
- --skip-go-mod-tidy flag to skip the tidy step (useful for CI or when the worker module is unreachable)

Static resources
- Embedded main.go template (minimal package main + empty main())
- Reuses existing .funcignore template
- New entry in StaticResources.props and StaticResources.cs

InitAction integration
- InitLanguageSpecificArtifacts handles WorkerRuntime.Native: calls GoHelpers.SetupProject, writes main.go and .funcignore
- Docker flag emits a warning: "Docker support for the native worker runtime is not yet available"
- Language selection wizard surfaces for Native runtime

Testing
- 27 unit tests passing (GoHelperTests + InitActionGoTests)
- 4 E2E test cases defined (NativeGoInitTests)
- Manual validation of 8 scenarios: --worker-runtime native/go/golang, --skip-go-mod-tidy, named subfolder, --force in non-empty dir, --docker warning, and interactive picker

| Test | Scenario                          | Result |
|------|-----------------------------------|--------|
| 1    | `--worker-runtime native`         | All files created; `FUNCTIONS_WORKER_RUNTIME: "native"` set; `go.mod` + `go.sum` present |
| 2    | `--worker-runtime go`             | Same output — alias works correctly |
| 3    | `--worker-runtime golang`         | Same output — alias works correctly |
| 4    | `--skip-go-mod-tidy`              | Shows “Skipped” message; `go.sum` still exists (from `go get`) |
| 5    | Named subfolder `test-go-5`        | Created subfolder and initialized inside it |
| 6    | `--force` in non-empty directory  | `dummy.txt` preserved alongside all new files |
| 7    | `--docker`                        | Docker warning emitted: “Docker support for the native worker runtime is not yet available. Skipping Dockerfile creation.” No Dockerfile created |
| 8    | Interactive (no runtime flag)     | Picker shows **Native** as option 7 in the list |

Output from local testing
[test-results.txt](https://github.com/user-attachments/files/26729550/test-results.txt)

### CODE REVIEW

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs [hunk 1]
```diff
@@ -276,7 +282,7 @@ private static (WorkerRuntime WorkerRuntime, string WorkerLanguage) ResolveWorke
                 workerRuntime = WorkerRuntimeLanguageHelper.NormalizeWorkerRuntime(workerRuntimeString);
                 language = !string.IsNullOrEmpty(languageString)
                     ? WorkerRuntimeLanguageHelper.NormalizeLanguage(languageString)
-                    : WorkerRuntimeLanguageHelper.NormalizeLanguage(workerRuntimeString);
+                    : WorkerRuntimeLanguageHelper.GetDefaultTemplateLanguageFromWorker(workerRuntime);
```
Comment (harshivcodes) [inline src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs]: this may have side effects for other languages, re-evaluate this chane

File: src/Cli/func/Helpers/WorkerRuntimeLanguageHelper.cs [hunk 2]
```diff
@@ -145,7 +153,16 @@ public static WorkerRuntime NormalizeWorkerRuntime(string workerRuntime)
...
             else if (_normalizeMap.ContainsKey(workerRuntime))
             {
-                return _normalizeMap[workerRuntime];
+                var resolved = _normalizeMap[workerRuntime];
+
+                if (resolved == WorkerRuntime.Native
+                    && !string.Equals(workerRuntime, "native", StringComparison.OrdinalIgnoreCase))
+                {
+                    ColoredConsole.WriteLine(WarningColor(
+                        $"'{workerRuntime}' is not a worker runtime. Golang is supported under the 'native' worker runtime. Resolving to --worker-runtime native."));
```
Comment (liliankasem) [inline src/Cli/func/Helpers/WorkerRuntimeLanguageHelper.cs]: This is technically correct but may confuse users who intentionally typed `go`. Consider rewording to something like:

> *"'go' maps to the 'native' worker runtime. Resolving to --worker-runtime native."*

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs [hunk 3]
```diff
@@ -369,6 +391,13 @@ private static async Task InitLanguageSpecificArtifacts(
                 case Helpers.WorkerRuntime.Node:
                     await NodeJSHelpers.SetupProject(programmingModel, language);
                     break;
+                case Helpers.WorkerRuntime.Native:
+                    await GoHelpers.SetupProject(
+                        Utilities.SanitizeLiteral(Path.GetFileName(Environment.CurrentDirectory), allowed: "-_."),
```
Comment (liliankasem) [inline src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs]: Go module names follow different conventions than typical sanitization (they're often URL-like paths, e.g., `github.com/org/repo`). Using the current directory name is reasonable for a default, but the restrictive character set could produce unexpected module names for directories with special characters. Worth verifying edge cases.

File: src/Cli/func/StaticResources/main.go [hunk 4]
```diff
@@ -0,0 +1,4 @@
+package main
+
+func main() {
+}
```
Comment (liliankasem) [inline src/Cli/func/StaticResources/main.go]: nit: add a new line to end of file, most Go tooling (gofmt, editors, linters) expects a trailing newline

```suggestion
func main() { }

```

File: src/Cli/func/Helpers/GoHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                    var match = Regex.Match(output, @"go(\d+\.\d+(?:\.\d+)?)");
+                    if (match.Success)
+                    {
+                        return new WorkerLanguageVersionInfo(WorkerRuntime.Native, match.Groups[1].Value, goExe);
+                    }
+                }
+            }
+            catch (Exception)
+            {
+                // Go is not installed or not on PATH
```
Comment (liliankasem) [inline src/Cli/func/Helpers/GoHelpers.cs]: Should we log to the user about this? Trying to understand why we would want to swallow an exception here?

File: src/Cli/func/Helpers/GoHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        public static async Task SetupProject(string moduleName, bool skipGoModTidy)
+        {
+            var goVersion = await GetEnvironmentGoVersion();
+            AssertGoVersion(goVersion);
+
+            // Initialize: Run go mod init
+            var modInitExe = new Executable("go", $"mod init {moduleName}");
+            var modInitExitCode = await modInitExe.RunAsync(
+                l => ColoredConsole.WriteLine(l),
+                e => ColoredConsole.Error.WriteLine(ErrorColor(e)));
```
Comment (liliankasem) [inline src/Cli/func/Helpers/GoHelpers.cs]: When I ran "func init --worker-runtime native" I noticed go text is red

```bash
go: creating new go.mod: module go-app

go: downloading github.com/azure/azure-functions-golang-worker v0.0.0-20260414214309-93dd9b8d6637

go: added github.com/azure/azure-functions-golang-worker v0.0.0-20260414214309-93dd9b8d6637

go: warning: "all" matched no packages
```

Given that we are setting the error color here, it appears that the Go stderr output (even informational) renders red. So I am guesing that `go get` and `go mod tidy` write progress to stderr instead of stdout, even though there isnt an error. The stderr handler should use plain output or AdditionalInfoColor for non-failure cases.

Second, we show suppress this information unless the global --verbose flag is passed e.g.

```csharp
 var modInitExitCode = modInitExe exe.RunAsync(
     l =>
     {
         if (GlobalCoreToolsSettings.IsVerbose)
         {
             ColoredConsole.WriteLine(VerboseColor(l));
         }
     },
     e =>
     {
         if (GlobalCoreToolsSettings.IsVerbose)
         {
             ColoredConsole.WriteLine(VerboseColor(e));
         }
     });
```

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs [hunk 7]
```diff
@@ -209,6 +215,11 @@ private async Task InitFunctionAppProject()
             {
                 (ResolvedWorkerRuntime, ResolvedLanguage) = ResolveWorkerRuntimeAndLanguage(WorkerRuntime, Language);
 
+                if (ResolvedWorkerRuntime == Helpers.WorkerRuntime.Native)
+                {
+                    ColoredConsole.WriteLine(WarningColor("[preview] The native worker runtime is currently in preview. Features and behavior may change in future releases."));
+                }
```
Comment (liliankasem) [inline src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs]: This shouldn't live in the init action, it should be written any time the worker runtime is `native` globally.

The existing pattern is `Utilities.WarnIfPreviewVersion()` in Utilities.cs, it's called from InitAction, StartHostAction, and HelpAction. That's the right place to add something like:

```
     if (GlobalCoreToolsSettings.CurrentWorkerRuntimeOrNone == WorkerRuntime.Native)
     {
         ColoredConsole.WriteLine(WarningColor(
             "The native worker runtime is currently in preview. ..."));
     }
```

(I would drop the [preview] pattern)

You can then call `Utilities.WarnIfPreviewVersion()` in new places if needed (i.e. pack action etc. later on)

File: src/Cli/func/Helpers/GoHelpers.cs (hunk 6 above)
Comment (liliankasem) [reply src/Cli/func/Helpers/GoHelpers.cs]: Might want this to be a helper method since it needs to apply to all commands (go mod init, go get, go mod tidy)

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs [hunk 8]
```diff
@@ -71,6 +71,8 @@ public InitAction(ITemplatesManager templatesManager, ISecretsManager secretsMan
```
Comment (liliankasem) [inline src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs:71]: Running `func init --worker-runtime native` should prompt for language. Right now it just assumes `go` by default.

`func init --worker-runtime native -l go` should be the path to full command run without prompts

File: src/Cli/func/StaticResources/main.go [hunk 9]
```diff
@@ -0,0 +1,4 @@
+package main
```
Comment (liliankasem) [inline src/Cli/func/StaticResources/main.go:1]: Does this need to be created by init? I would expect this to be added by `func new` as an actual http template

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs [hunk 10]
```diff
@@ -369,6 +391,13 @@ private static async Task InitLanguageSpecificArtifacts(
                 case Helpers.WorkerRuntime.Node:
                     await NodeJSHelpers.SetupProject(programmingModel, language);
                     break;
+                case Helpers.WorkerRuntime.Native:
+                    await GoHelpers.SetupProject(
+                        Utilities.SanitizeLiteral(Path.GetFileName(Environment.CurrentDirectory), allowed: "-_."),
+                        skipGoModTidy);
+                    await FileSystemHelpers.WriteFileIfNotExists("main.go", await StaticResources.MainGo);
+                    await FileSystemHelpers.WriteFileIfNotExists(Constants.FuncIgnoreFile, await StaticResources.FuncIgnore);
```
Comment (liliankasem) [inline src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs]: I would expect all these operations to be part of `GoHelpers.SetupProject`

File: src/Cli/func/Actions/LocalActions/InitAction/InitNativeSubcommandAction.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+namespace Azure.Functions.Cli.Actions.LocalActions
+{
+    [Action(Name = "init native", ParentCommandName = "init", ShowInHelp = true, HelpText = "Options specific to native runtime apps when running func init")]
+    internal class InitNativeSubcommandAction : BaseAction
+    {
+        public override ICommandLineParserResult ParseArgs(string[] args)
+        {
+            Parser
+                .Setup<string>('l', "language")
+                .WithDescription("The language for the function app. Options: golang.")
```
Comment (liliankasem) [inline src/Cli/func/Actions/LocalActions/InitAction/InitNativeSubcommandAction.cs]: nit: I know not all of the other commands are doing this, but this should ideally not be hardcoded. Can do similar to how we get the tfm for dotnet, but for supported native languages

```csharp
            Parser
                .Setup<string>("target-framework")
                .WithDescription($"The target framework for the .NET project. Options: {string.Join(", ", TargetFrameworkHelper.GetSupportedTargetFrameworks())}.")
                .Callback(_ => { });
```

File: src/Cli/func/Actions/LocalActions/InitAction/InitNativeSubcommandAction.cs [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+                .Callback(_ => { });
+
+            Parser
+                .Setup<bool>("skip-go-mod-tidy")
+                .WithDescription("Skip running 'go mod tidy' after project creation.")
```
Comment (liliankasem) [inline src/Cli/func/Actions/LocalActions/InitAction/InitNativeSubcommandAction.cs]: nit 
```suggestion
                .WithDescription("Skip running 'go mod tidy' after Go project creation.")
```

File: src/Cli/func/Common/Constants.cs [hunk 13]
```diff
@@ -113,7 +113,8 @@ internal static partial class Constants
             { WorkerRuntime.Dotnet, new[] { "mcr.microsoft.com/azure-functions/dotnet", "microsoft/azure-functions-dotnet-core2.0", "mcr.microsoft.com/azure-functions/base", "microsoft/azure-functions-base" } },
             { WorkerRuntime.Node, new[] { "mcr.microsoft.com/azure-functions/node", "microsoft/azure-functions-node8" } },
             { WorkerRuntime.Python, new[] { "mcr.microsoft.com/azure-functions/python", "microsoft/azure-functions-python3.6" } },
-            { WorkerRuntime.Powershell, new[] { "mcr.microsoft.com/azure-functions/powershell", "microsoft/azure-functions-powershell" } }
+            { WorkerRuntime.Powershell, new[] { "mcr.microsoft.com/azure-functions/powershell", "microsoft/azure-functions-powershell" } },
+            { WorkerRuntime.Native, new string[] { } }
```
Comment (liliankasem) [inline src/Cli/func/Common/Constants.cs]: Do we need to add this here? Were you running into any issues without it?

File: src/Cli/func/Helpers/GoHelpers.cs [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+            {
+                ColoredConsole.WriteLine(AdditionalInfoColor("Skipped \"go mod tidy\". You must run \"go mod tidy\" manually."));
+            }
+        }
+
+        private static async Task<WorkerLanguageVersionInfo> GetVersion(string goExe)
+        {
+            try
+            {
+                var exe = new Executable(goExe, "version");
```
Comment (liliankasem) [inline src/Cli/func/Helpers/GoHelpers.cs:119]: This pattern repeats itself multiple times. A helper like RunGoCommandAsync would clean this up:

```csharp
private static async Task RunGoCommandAsync(string arguments, string errorMessage, bool 
throwOnFailure = true)
 {
     var exe = new Executable("go", arguments);
     var exitCode = await exe.RunAsync(
         l =>
         {
             if (GlobalCoreToolsSettings.IsVerbose)
             {
                 ColoredConsole.WriteLine(VerboseColor(l));
             }
         },
         e =>
         {
             if (GlobalCoreToolsSettings.IsVerbose)
             {
                 ColoredConsole.WriteLine(VerboseColor(e));
             }
         });
 
     if (exitCode != 0)
     {
         if (throwOnFailure)
         {
             throw new CliException($"{errorMessage} 'go {arguments}' exited with code 
{exitCode}.");
         }
 
         ColoredConsole.WriteLine(WarningColor($"Warning: 'go {arguments}' exited with a 
non-zero code. You may need to run it manually."));
     }
 }
```

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs [hunk 15]
```diff
@@ -238,7 +249,7 @@ private async Task InitFunctionAppProject()
             else
             {
                 bool managedDependenciesOption = ResolveManagedDependencies(ResolvedWorkerRuntime, ManagedDependencies);
-                await InitLanguageSpecificArtifacts(ResolvedWorkerRuntime, ResolvedLanguage, ResolvedProgrammingModel, managedDependenciesOption, GeneratePythonDocumentation);
+                await InitLanguageSpecificArtifacts(ResolvedWorkerRuntime, ResolvedLanguage, ResolvedProgrammingModel, managedDependenciesOption, GeneratePythonDocumentation, SkipGoModTidy);
```
Comment (liliankasem) [inline src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs]: This is a bit of a code smell, and it started with `generatePythonDocumentation` to be frank.

Every runtime-specific flag gets bolted onto this shared method. generatePythonDocumentation was already a smell, and now skipGoModTidy makes it worse. If Rust or another native language lands later, it'll grow again.

I mention this in another comment, but the solution here is to move the file writes into GoHelpers.SetupProject (matching Node's pattern), then the Native case in the switch becomes just:

```csharp
   case Helpers.WorkerRuntime.Native:
       await GoHelpers.SetupProject(
           Utilities.SanitizeLiteral(...), skipGoModTidy);
       break;
```

And `skipGoModTidy` doesn't need to be a parameter of InitLanguageSpecificArtifacts at all, it can stay on the InitAction class and be passed directly. 

I.e. 

 // In InitFunctionAppProject(), before the shared method:
 if (ResolvedWorkerRuntime == WorkerRuntime.Native)
 {
     await GoHelpers.SetupProject(moduleName, SkipGoModTidy);
 }
 else
 {
     await InitLanguageSpecificArtifacts(ResolvedWorkerRuntime, ...);
 }

or use directly via 

this.SkipGoModTidy

File: test/Cli/Func.E2ETests/Commands/FuncInit/NativeGoInitTests.cs [hunk 16]
```diff
@@ -0,0 +1,95 @@
+// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (liliankasem) [inline test/Cli/Func.E2ETests/Commands/FuncInit/NativeGoInitTests.cs:1]: Missing test cases:

   1. No E2E test for --worker-runtime go alias — there's a unit test for 
  NormalizeWorkerRuntime("go") resolving to Native, but no E2E test verifying func init 
  --worker-runtime go actually works end-to-end and shows the redirect warning.
   2. No E2E test for init in non-empty directory without --force — there's a test with 
  --force, but not one asserting it fails/warns without it (other runtimes test this).
   3. No test for host.json content — Node tests verify host.json contains expected logging 
  config. The Go E2E checks host.json has extensionBundle but doesn't verify logging config.
   4. No test for local.settings.json having FUNCTIONS_WORKER_RUNTIME set to native — it's 
  validated in filesToValidate but only checks the strings exist, not the actual JSON 
  structure. This matches the existing pattern though, so not a regression.
   5. No unit test for GetVersion parsing edge cases — e.g., what happens with unexpected go 
  version output formats, or Go beta versions like go1.25beta1.

File: src/Cli/func/Helpers/GoHelpers.cs (hunk 5 above)
Comment (harshivcodes) [reply src/Cli/func/Helpers/GoHelpers.cs]: Here's the flow:
1. GetVersion catches the exception → returns null
2. SetupProject calls AssertGoVersion(null)
3. User sees: "Could not find a Go installation. Go 1.24 or later is required. Please install Go from https://go.dev/dl/"

So the user-facing message is good and actionable. The swallowed exception just hides why it failed (e.g., permission error vs not on PATH). I can log this exception if they are running in verbose mode.

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs (hunk 3 above)
Comment (harshivcodes) [reply src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs]: Good point, what appears in the go.mod file could turn into the executable name (unless the customer overrides it). I removed `.`  since that's the only potentially problematic character in the executable name. I think the others are ok.

File: src/Cli/func/StaticResources/main.go (hunk 9 above)
Comment (harshivcodes) [reply src/Cli/func/StaticResources/main.go:1]: Is this behavior I can change later or are we locked into it? If we can change it, I am inclined to leave it here since it's the equivalent of Python's function_app.py — a mandatory entry point, not a function template.

We are likely not going to add func new support for golang during preview - it might come slightly later.

Comment (liliankasem) [reply src/Cli/func/StaticResources/main.go:1]: Yeah that's fine with me

File: src/Cli/func/Helpers/GoHelpers.cs (hunk 6 above)
Comment (harshivcodes) [reply src/Cli/func/Helpers/GoHelpers.cs]: Done

File: src/Cli/func/Common/Constants.cs (hunk 13 above)
Comment (harshivcodes) [reply src/Cli/func/Common/Constants.cs]: removed it.

File: src/Cli/func/Helpers/GoHelpers.cs (hunk 14 above)
Comment (harshivcodes) [reply src/Cli/func/Helpers/GoHelpers.cs:119]: done.

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs (hunk 7 above)
Comment (harshivcodes) [reply src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs]: Made this change.

File: src/Cli/func/Actions/LocalActions/InitAction/InitNativeSubcommandAction.cs (hunk 11 above)
Comment (harshivcodes) [reply src/Cli/func/Actions/LocalActions/InitAction/InitNativeSubcommandAction.cs]: Updated.

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs (hunk 15 above)
Comment (harshivcodes) [reply src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs]: 1. Pulled Native out of the shared InitLanguageSpecificArtifacts method into its own block in InitFunctionAppProject
 2. Removed skipGoModTidy from the InitLanguageSpecificArtifacts signature
 3. GoHelpers.SetupProject is called directly with the flag

File: test/Cli/Func.E2ETests/Commands/FuncInit/NativeGoInitTests.cs (hunk 16 above)
Comment (harshivcodes) [reply test/Cli/Func.E2ETests/Commands/FuncInit/NativeGoInitTests.cs:1]: Added these tests.

File: src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs (hunk 8 above)
Comment (harshivcodes) [reply src/Cli/func/Actions/LocalActions/InitAction/InitAction.cs:71]: fixed.

**Review by liliankasem — APPROVED** [review]
